### PR TITLE
Rules profiling v1

### PR DESCRIPTION
**Pull request by regit**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/5667

Describe changes:
- rules only profiling
- unix socket activation and dump

### CODE REVIEW

**Comment by regit** [comment]
I need to add the documentation.

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW1_stats_chk | | |
| .tcp.rst | 128237 | 104135 | 81.21% | 

Pipeline 10576

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW1_stats_chk | | |
| .tcp.rst | 128237 | 103902 | 81.02% | 

Pipeline 10579

File: doc/userguide/rule-management/rule-profiling.rst [hunk 1]
```diff
@@ -0,0 +1,28 @@
+Rules Profiling
+===============
+
+If Suricata is built with the --enable-profiling-rules then the ruleset profiling
```
Comment (jufajardini) [inline doc/userguide/rule-management/rule-profiling.rst:4]: (if there's a new version for this work)
nit: highlight the option more:
"with the `--enable-profiling-rules` option..."

File: doc/userguide/rule-management/rule-profiling.rst [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+A typical scenario to get rules performance would be ::
+
+ surictasc -c ruleset-profile-start
+ sleep 30
+ surictasc -c ruleset-profile-stop
+ suricatasc -c ruleset-profile
+
+On busy systems, using the sampling capability to capture performance
+on a subset of packets can be obtained via the `sample-rate` variable
+in the `profiling` section.
```
Comment (jufajardini) [inline doc/userguide/rule-management/rule-profiling.rst:28]: Nit, only if there's a new version for this PR:
"in the `profiling` section in the suricata.yaml file."

**Review by jufajardini — APPROVED** [review]
Doc looking good to me. IFF there's a new version for this work, then I left two nit comments :P

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on SURI_TLPW1_stats_chk.
  

Pipeline 10579

**Comment by victorjulien** [comment]
I've rebased this and will push a new PR.

File: src/util-profiling.c [hunk 3]
```diff
@@ -1399,4 +1399,76 @@ SCProfilingRegisterTests(void)
...
+    SCReturnInt(TM_ECODE_OK);
+}
+
+#elif PROFILE_RULES
+
+thread_local int profiling_rules_entered = 0;
+int profiling_output_to_file = 0;
+static SC_ATOMIC_DECLARE(uint64_t, samples);
+static uint64_t rate = 0;
+int profiling_rules_active = 0;
```
Comment (victorjulien) [inline src/util-profiling.c:1418]: bool

File: src/util-profiling.c [hunk 4]
```diff
@@ -1399,4 +1399,76 @@ SCProfilingRegisterTests(void)
...
+        p->flags |= PKT_PROFILE;
+        return 1;
+    }
+
+    if (p->flags & PKT_PROFILE)
+        return 1;
+    return 0;
+}
+
+int SCProfileRuleStartCollection(void)
```
Comment (victorjulien) [inline src/util-profiling.c:1462]: unclear why we return int that is always the same value

File: src/util-profiling.h [hunk 5]
```diff
@@ -387,4 +364,72 @@ void SCProfilingDump(void);
...
+int SCProfileRuleStopCollection(void);
+void SCProfilingRuleThreatAggregate(DetectEngineThreadCtx *det_ctx);
+
+#define RULE_PROFILING_START(p)                                                                    \
+    uint64_t profile_rule_start_ = 0;                                                              \
+    uint64_t profile_rule_end_ = 0;                                                                \
+    if (profiling_rules_enabled && SCProfileRuleStart((p))) {                                      \
+        if (profiling_rules_entered > 0) {                                                         \
+            SCLogError(SC_ERR_FATAL, "Re-entered profiling, exiting.");                            \
+            exit(1);                                                                               \
```
Comment (victorjulien) [inline src/util-profiling.h:414]: SCFatal

File: src/util-profiling.h [hunk 6]
```diff
@@ -387,4 +364,72 @@ void SCProfilingDump(void);
...
+        profiling_rules_entered++;                                                                 \
+        profile_rule_start_ = UtilCpuGetTicks();                                                   \
+    }
+
+#define RULE_PROFILING_END(ctx, r, m, p)                                                           \
+    if (profiling_rules_enabled && ((p)->flags & PKT_PROFILE)) {                                   \
+        profile_rule_end_ = UtilCpuGetTicks();                                                     \
+        SCProfilingRuleUpdateCounter(                                                              \
+                ctx, r->profiling_id, profile_rule_end_ - profile_rule_start_, m);                 \
+        profiling_rules_entered--;                                                                 \
```
Comment (victorjulien) [inline src/util-profiling.h:425]: lets put a overflow check (BUG_ON) here

File: src/util-profiling.c [hunk 7]
```diff
@@ -1399,4 +1399,76 @@ SCProfilingRegisterTests(void)
...
+{
+    profiling_rules_active = 1;
```
Comment (victorjulien) [inline src/util-profiling.c:1464]: global, shared between threads, so should be atomic

File: src/counters.c [hunk 8]
```diff
@@ -511,6 +511,7 @@ static void *StatsWakeupThread(void *arg)
             /* assuming the assignment of an int to be atomic, and even if it's
              * not, it should be okay */
             tv->perf_public_ctx.perf_flag = 1;
+            tv->profile_flag = 1;
```
Comment (victorjulien) [inline src/counters.c:514]: I don't like how the detect specific logic flows into the core threading code here

File: src/unix-manager.c [hunk 9]
```diff
@@ -793,6 +794,49 @@ static TmEcode UnixManagerRulesetStatsCommand(json_t *cmd,
...
+    json_object_set_new(server_msg, "message", js);
+    SCReturnInt(TM_ECODE_OK);
+}
+
+static TmEcode UnixManagerRulesetProfileStartCommand(json_t *cmd, json_t *server_msg, void *data)
+{
+    SCEnter();
+
+    int ret = SCProfileRuleStartCollection();
+    if (ret != TM_ECODE_OK) {
```
Comment (victorjulien) [inline src/unix-manager.c:817]: never true

File: src/unix-manager.c [hunk 10]
```diff
@@ -793,6 +794,49 @@ static TmEcode UnixManagerRulesetStatsCommand(json_t *cmd,
...
+    json_object_set_new(server_msg, "message", json_string("OK"));
+    SCReturnInt(TM_ECODE_OK);
+}
+
+static TmEcode UnixManagerRulesetProfileStopCommand(json_t *cmd, json_t *server_msg, void *data)
+{
+    SCEnter();
+
+    int ret = SCProfileRuleStopCollection();
+    if (ret != TM_ECODE_OK) {
```
Comment (victorjulien) [inline src/unix-manager.c:830]: never true

File: src/counters.c (hunk 8 above)
Comment (victorjulien) [reply src/counters.c:514]: @regit what is the goal here? To sync threads on rule profile data at the same interval as normal stats? Or is the frequency not very important?

**Comment by victorjulien** [comment]
replaced by #8871

**Pull request closed without merging**